subsys/bluetooth/host/conn.c

```diff
@@ -1823,6 +1824,20 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 	return -EINVAL;
 }
 
+#if defined(CONFIG_BT_TESTING)
+int bt_conn_remote_info(const struct bt_conn *conn, struct bt_conn_rv_info *rv)
+{
+	if (conn->type == BT_CONN_TYPE_LE) {
```

Comment:
Sorry but this is not quite right, the idea was that hci_read_remote_version would be called, if it hasn't been called yet, so that only application using this function would incur in extra latency fetching this info. Also that means you don't need to hide this functionality behing CONFIG_BT_TESTING.

Reply:
Fair point, but that is above my skill level.

Reply:
Not sure I understand you comment about skill level, what I suggested here is to check if rv has been read and if not call hci_read_remote_version which should use bt_hci_cmd_send_sync and block until the info is returned.

Reply:
What I mean is that I cannot make something that both does the task it is supposed to do and still follow the code standards.

I could for example use something like:
if (conn->rv == null) {
    hci_read_remote_version(conn);
    while (conn->rv == null) { /* or something with a timeout */
        k_sleep(/* some reasonable time */);
    }
} else {
    /* assign the values */
}

But I have a strong suspicion that this is not up to the standards set.

Reply:
Just make hci_read_remote_version use bt_hci_cmd_send_sync that one will block waiting the response on semaphore so you don't need to do any busy loop or anything:

https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L307

So all you need to do is replace bt_hci_cmd_send with bt_hci_cmd_send_sync, parse the response setting rv.

Reply:
I pushed the latest changes. As you mentioned the CONFIG_BT_TESTING would not be needed, I wondered if it should be wrapped in a broader config like CONFIG_BT_CONN or something.

Reply:
Yes, if it depends on things like bt_conn it should be conditional to CONFIG_BT_CONN if that is not already the case, but it looks like this is still not calling hci_read_remote_version directly so the change to bt_hci_cmd_send_sync is useless as the processing is asynchronous which is racy as the event may come after bt_conn_get_info is called thus the suggestion to use bt_hci_cmd_send_sync and process the event synchronously.

Reply:
Let me see if I get this right. bt_hci_cmd_send_sync only blocks until the command is executed, but does not wait for the event?

Reply:
bt_hci_cmd_send_sync only works with commands that respond via command complete, if the response comes as a separate event then we either have to have it to work with an arbritrary opcode or have a separate semaphore to block on until we wait the event to arrive.

Reply:
Do you have a suggestion how to share a semaphore between hci_read_remote_version and bt_hci_evt_read_remote_version_complete?

Reply:
Implemented with callback instead. Tested with my application and it worked.